### PR TITLE
Lift manifest name restrictions on `swift package edit`

### PR DESCRIPTION
**Pull request by stevapple**
Lift package manifest name restrictions on `swift package edit`, since it's already replaced with package identity.

### Motivation:

Package manifest name no longer has actual usage, and it is replaced by a case-insensitive package identity. There's no reason to forbid users from editing a package with a different manifest name than expected identity.

### Modifications:

- Switch to case-insensitive comparison between manifest name and package identity;
- Emit a warning instead of an error if these two don't match.

### Result:

`swift package edit` will get far fewer restrictions, making it more usable.

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]
Seems reasonable to me!

File: Sources/Workspace/Workspace.swift [hunk 1]
```diff
@@ -1373,8 +1373,8 @@ extension Workspace {
                                   completion: $0)
             }
 
-            guard manifest.displayName == packageName else {
-                return observabilityScope.emit(error: "package at '\(destination)' is \(manifest.displayName) but was expecting \(packageName)")
+            if manifest.displayName.lowercased() != packageName.lowercased() {
+                observabilityScope.emit(warning: "package at '\(destination)' is \(manifest.displayName) but was expecting \(packageName)")
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:1377]: should we add a test case?

**Comment by abertelrud** [comment]
@stevapple Would it be possible to add a test case here so we can merge it?

**Comment by neonichu** [comment]
@stevapple Do you need any help with adding a test for this?

**Comment by stevapple** [comment]
Sorry for the late response — I tried today and found that `MockManifestLoader` falsely shadows the edited package so we need to tweak the mock system first. Hopefully we can work around the problem.

**Comment by stevapple** [comment]
cc @tomerd @neonichu @abertelrud

File: Sources/Workspace/Workspace.swift (hunk 1 above)
Comment (stevapple) [reply Sources/Workspace/Workspace.swift:1382]: done👌

File: Sources/SPMTestSupport/MockManifestLoader.swift [hunk 2]
```diff
@@ -63,11 +63,16 @@ public final class MockManifestLoader: ManifestLoaderProtocol {
...
+            let possiblyEditedManifestKey = Key(url: manifestPath.parentDirectory.pathString, version: nil)
+
+            if possiblyEditedManifestKey != manifestKey && packageVersion == nil,
+                let result = self.manifests[possiblyEditedManifestKey] {
+                return completion(.success(result))
+            } else if let result = self.manifests[manifestKey] {
                 return completion(.success(result))
             } else {
-                return completion(.failure(MockManifestLoaderError.unknownRequest("\(key)")))
+                return completion(.failure(MockManifestLoaderError.unknownRequest("\(manifestKey)")))
```
Comment (tomerd) [inline Sources/SPMTestSupport/MockManifestLoader.swift:75]: could you elaborate on why this change was needed? it seems a bit fragile

Comment (stevapple) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: What do you mean by “this change” — the introduction of `possiblyEditedManifestKey`, or the renaming of `key` to `manifestKey`?

Comment (tomerd) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: sorry for the confusing language. I meant the introduction of possiblyEditedManifestKey and related logic

Comment (stevapple) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: When we edit a package to a local path, SwiftPM will shadow it with a new `packagePath` but keeps `packageLocation` unchanged to preserve its identity. The `packagePath` is not passed directly to `ManifestLoader` — instead, a `manifestPath` is computed from `packagePath`, which indicates the new manifest location to read.

When using a real `Workspace` this works as expected, because `ManifestLoader` will always execute and read from the edited manifest. The `MockManifestLoader`, however, only takes `packageLocation` and `packageVersion` into account. `packageLocation` is unchanged before and after edits, so editing a `MockWorkspace` don’t actually do anything to the manifest.

`possiblyEditedManifestKey` is the key of the package from a possibly edited manifest path. Such edited package should:
- Have a different `packagePath.pathString` from `packageLocation`;
- Satisfy `packageVersion == nil`;
- Ensure `manifestPath` is exactly the `Package.swift` file located in `packagePath` directory.

So, when `possiblyEditedManifestKey != manifestKey && packageVersion == nil`, the manifest is likely an edited one from other package, and we should load it using `possiblyEditedManifestKey`. In fact, we can remove `possiblyEditedManifestKey != manifestKey` because when the two is the same the lookup still works. I kept the judgement because this allows us to set breakpoints separately for edited and unedited cases.

The only possible exception here is when we use URL or registry for a package without specifying its revision or version. It’s possible for directly using a `MockManifestLoader`, but that’s an anti-pattern and we won’t touch it when using wrapped functionalities provided by `MockWorkspace`, so I believe that’s fine.

Comment (tomerd) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: I see. is there a different way we can implement this such that its more clear instead of a heuristic? for example, can we add some more information to what MockManifestLoader uses such that it supports the edit use case better?

Comment (stevapple) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: I’m afraid it’s hard to do so because the interface (and even implementation) is from `ManifestLoaderProtocol`, and modifying this protocol will:
- cause a large scale of codes to be updated unnecessarily;
- have an impact on production codes instead of resolving it within `SPMTestSupport`;
- introduce a useless  (and somehow confusing) new argument that may have unexpected consequences like being misused.

Even if source compatibility of `libSwiftPM` is not a goal for now, interface changes should be avoided when possible — not to say that this one is almost unnecessary.

Comment (stevapple) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: I’ve tried my best to make the code self-explain. If you think additional notes here are necessary, just tell me what to add:)

Comment (tomerd) [reply Sources/SPMTestSupport/MockManifestLoader.swift:75]: the issue with the change proposed in this PR is that it is a heuristics that is likely to cause test reliability issues down the line. 

I dont think we should update ManifestLoaderProtocol. you should be able to inject a custom manifest loader in the tests and then operate on the concrete mock type, ie you dont have to even use MockManifestLoader

**Comment by neonichu** [comment]
Closing due to inactivity.
